CreateSampleWorkspace to also allow creating monitors

ianbush opened the pull request:
This adds the option "NumMonitors" to CreateSampleWorkspace.

**To test:**

* Run CreateSampleWorkspace with the NumMonitors and any a variety of other options, checking the monitors are added to the workspace (their workspace indices will be the after the detectors)
* Check the monitors are visible in the instrument view
* Check the updated documentation looks good

Fixes #17477 .

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

KarlPalmen commented on Framework/Algorithms/inc/MantidAlgorithms/CreateSampleWorkspace.h:
```diff
@@ -63,8 +63,9 @@ class DLLExport CreateSampleWorkspace : public API::Algorithm {
   Geometry::Instrument_sptr createTestInstrumentRectangular(
-      API::Progress &progress, int num_banks, int pixels, double pixelSpacing,
-      const double bankDistanceFromSample, const double sourceSampleDistance);
+      API::Progress &progress, int num_banks, int numMonitors, int pixels,
```
``numMonitors`` is not consistent with ``num_banks``.

KarlPalmen commented on Framework/Algorithms/src/CreateSampleWorkspace.cpp:
```diff
@@ -497,6 +504,35 @@ Instrument_sptr CreateSampleWorkspace::createTestInstrumentRectangular(
+    bank->initialize(pixelShape, 1, 0.0, pixelSpacing, 1, 0.0, pixelSpacing,
+                     monitorNumber, true, 1);
+
+    // Mark them all as detectors
```
It looks like you are marking them all as monitors.

KarlPalmen commented on Framework/Algorithms/test/CreateSampleWorkspaceTest.h:
```diff
@@ -452,6 +452,63 @@ class CreateSampleWorkspaceTest : public CxxTest::TestSuite {
+
+    // Remove workspace from the data service.
+    AnalysisDataService::Instance().remove("outWS");
+  }
```
You set the number of monitors as 2, but test for 3 monitors. Perhaps, you need to test that the 3rd detector is not a monitor.

KarlPalmen replied:
Now I see that the first detector is tested to not be a monitor. I expected this to be the 3rd.

KarlPalmen commented:
I found that the monitors were given large spectra numbers than the other detectors, which other workspaces I've seen with monitors have the smallest spectra numbers given to the monitors.

ianbush commented:
I have now changed this so the monitors get the first spectrum numbers in the workspace. We have reversed this for some ILL instruments, but I accept this is not normal!

I have also fixed the comments and the inconsistent variable names.

ianbush commented:
This is causing a crash in the SmoothNeighburs doc test, I will take a look as to why.

KarlPalmen commented:
When looking at the instrument view, I found names like "Bank1", "Bank2", "Monitor3", "Monitor4". Would it be more appropriate to have something like "Bank1", "Bank2", "Montitor1", Monitor2"?

ianbush commented:
Yes, I had not noticed that, it is now fixed so the monitors get the numbers you would expect.

KarlPalmen commented:
:shipit:
